### PR TITLE
Java 21

### PR DESCRIPTION
**Pull request by TwoOfTwelve**
To update java for java 21 the following things have to be done:
- [x] Update java version in pom.xml
- [x] Update java version in workflow files
- [x] Update expected java version for tests
- [x] Verify the old tests are still working
- [x] Write rules and tests for new java features
- [x] Verify the rules below indeed do not impact jplag
- [x] Enable Previews
- [x] Right now the java version for enabling previews is hard coded. It should be unified with the version in NewJavaFeaturesTest

The following new features are relevant to jplag:
- [x] Record Patterns and Pattern matching for switch have been finalized
    Works without changes
- [x] String templates (preview)
    Works without changes
- [x] Unnamed patterns and variables (preview)
    Ignore vardefs without name (javac maps "_" to an empty string)
- [x] Anonymous classes. Produced negative positions for end of class tokens
- [x] Guards for cases. Should be extracted as an additional if.

The following changes should not impact jplag:
- [x] Virtual threads / Structured concurrency
- [x] Foreign function api (preview) [maybe poses problems for obfuscation of plagiarisms]
- [x] Scoped values

### CODE REVIEW

**Comment by SirYwell** [comment]
There are a few things that aren't really covered:

- case pattern with guards (`case Integer i when i > 0:`)
  - it might make sense to desugar it into an `if`
- [JEP 445](https://openjdk.org/jeps/445) (though it might make sense to skip that, there are a few changes to how the compiler handles it in Java 22)

Is that intentional?

File: languages/java/src/main/java/de/jplag/java/TokenGeneratingTreeScanner.java [hunk 1]
```diff
@@ -282,7 +282,23 @@ public Void visitSwitchExpression(SwitchExpressionTree node, Void unused) {
...
+        }
+        this.scan(node.getGuard(), null);
+        if (node.getCaseKind() == CaseTree.CaseKind.RULE) {
+            this.scan(node.getBody(), null);
+        } else {
+            this.scan(node.getStatements(), null);
+        }
+
+        if (node.getGuard() != null) {
+            addToken(JavaTokenType.J_IF_BEGIN, positions.getEndPosition(ast, node), 0, CodeSemantics.createControl());
```
Comment (SirYwell) [inline languages/java/src/main/java/de/jplag/java/TokenGeneratingTreeScanner.java:298]: This should be `END` I guess?

**Review by tsaglam — APPROVED** [review]

**Review by Kr0nox — CHANGES_REQUESTED** [review]
Just one small thing

File: languages/java/src/main/java/de/jplag/java/TokenGeneratingTreeScanner.java [hunk 2]
```diff
@@ -444,22 +460,24 @@ public Void visitAssert(AssertTree node, Void unused) {
...
-        boolean inLocalScope = variableRegistry.inLocalScope();
-        // this presents a problem when classes are declared in local scopes, which can happen in ad-hoc implementations
-        CodeSemantics semantics;
-        if (inLocalScope) {
-            boolean mutable = isMutable(node.getType());
-            variableRegistry.registerVariable(name, VariableScope.LOCAL, mutable);
-            semantics = new CodeSemantics();
-        } else {
-            semantics = CodeSemantics.createKeep();
+        if (!node.getName().contentEquals("")) {
```
[the remark this answers is not in the document]
Comment (tsaglam) [reply languages/java/src/main/java/de/jplag/java/TokenGeneratingTreeScanner.java:463]: Good call, I think we even talked about that before!

**Review by Kr0nox — APPROVED** [review]

**Comment by dfuchss** [comment]
-> https://github.com/jplag/JPlag/actions/runs/7057421814/workflow#L54

Missed one last line :)

**Pull request merged**